# 4.9 Task _Part 1

## Q3: Import your analysis libraries, as well as your new customer data set as a dataframe

In [8]:
## Import Libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [7]:
# Import ords_prods_merge Data set 

ords_prods_merge =pd.read_pickle(r'C:\Users\inesh\Instacart Basket Analysis_01_02_2024\02 Data\Prepared Data\ords_prods_merge_loyalty_spending_frequency_updated.pkl')

In [9]:
## Import customer data set from Original data

df_customer_original =pd.read_csv(r'C:\Users\inesh\Instacart Basket Analysis_01_02_2024\02 Data\Original Data\customers.csv', index_col = False)

In [4]:
df_customer_original.shape

(206209, 10)

In [5]:
df_customer_original.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

## Q4: Wrangle df_customer_original data 

#### Change column names & change into small case

In [14]:
df_customer_original.rename(columns={'First Name':'first_name',
                                     'Surnam':'last_name',
                                     'Gender':'gender',
                                     'STATE':'state',
                                     'Age':'age',
                                     'n_dependants':'dependant_no',
                                     'fam_status':'family_status'},
                            inplace=True)

In [7]:
df_customer_original.columns

Index(['user_id', 'first_name', 'last_name', 'gender', 'state', 'age',
       'date_joined', 'dependant_no', 'family_status', 'income'],
      dtype='object')

#### Check the data types of all the columns

In [8]:
df_customer_original.dtypes

user_id           int64
first_name       object
last_name        object
gender           object
state            object
age               int64
date_joined      object
dependant_no      int64
family_status    object
income            int64
dtype: object

#### Convert the data type of 'user_id' from integer to string (id fields is not used for any numerical calculations)


In [9]:
df_customer_original['user_id']=df_customer_original['user_id'].astype('str')

In [10]:
df_customer_original.dtypes

user_id          object
first_name       object
last_name        object
gender           object
state            object
age               int64
date_joined      object
dependant_no      int64
family_status    object
income            int64
dtype: object

In [11]:
df_customer_original.describe()

age   dependant_no         income
count  206209.000000  206209.000000  206209.000000
mean       49.501646       1.499823   94632.852548
std        18.480962       1.118433   42473.786988
min        18.000000       0.000000   25903.000000
25%        33.000000       0.000000   59874.000000
50%        49.000000       1.000000   93547.000000
75%        66.000000       3.000000  124244.000000
max        81.000000       3.000000  593901.000000

## Q5: Complete the fundamental data quality and consistency checks

### Check for missing values

In [13]:
df_customer_original.isnull().sum()

user_id              0
first_name       11259
last_name            0
gender               0
state                0
age                  0
date_joined          0
dependant_no         0
family_status        0
income               0
dtype: int64

#### There are  11259 missing values in first_name column. No need to delete or replace this values as we can use user_is as unique column of customer data

### Check for Duplicate values

In [10]:
# Find duplicates in the DataFrame based on all columns

duplicate_rows = df_customer_original[df_customer_original.duplicated()]

In [11]:
print(duplicate_rows)

Empty DataFrame
Columns: [user_id, First Name, Surnam, Gender, STATE, Age, date_joined, n_dependants, fam_status, income]
Index: []


### There are no duplicate rows in df_customer_original

#### Check for Mix-data types

In [12]:
# check for columns that have mixed data type

for col in df_customer_original.columns.tolist():
    weird = (df_customer_original[[col]].map(type) != df_customer_original[[col]].iloc[0].apply(type)).any(axis = 1)
    
    if len (df_customer_original[weird]) > 0:
        print (col)

First Name


First name column has mix  data type

In [15]:
# changing data type
df_customer_original['first_name'] = df_customer_original['first_name'].astype('str')

In [16]:
df_customer_original['first_name'].dtype

dtype('O')

## Q6: Combine your customer data with the rest of your prepared Instacart data

In [17]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_from_last_order  product_id  add_to_cart_order  reordered _merge  ...  \
0                   NaN         196                  1          0   both  ...   
1                  15.0         196                  1          1   both  ...   
2                  21.0         196                  1          1   both  ...   
3                  29.0         196                  1          1   both  ...   
4                  28.0         196                  1          1   both  ...   

     price_range_loc     busiest_day  busiest_days  busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regular days         Average_orders   
1  Mid-range product  Regularly busy  Slowest days          Fewest_orders   
2  Mid-range product  Regularly busy  Slowest days            Most_orders   
3  Mid-range product      Least busy  Slowest days          Fewest_orders   
4  Mid-range product      Least busy  Slowest days            Most_orders   

   max_order  loyalty_flag average_spending spending flag  \
0         10  New customer         6.367797   Low spender   
1         10  New customer         6.367797   Low spender   
2         10  New customer         6.367797   Low spender   
3         10  New customer         6.367797   Low spender   
4         10  New customer         6.367797   Low spender   

  frequency_of_purchase   order frequency flag  
0                  20.5  Non-frequent customer  
1                  20.5  Non-frequent customer  
2                  20.5  Non-frequent customer  
3                  20.5  Non-frequent customer  
4                  20.5  Non-frequent customer  

[5 rows x 26 columns]

#### Drop merge column from ords_prod dataframe to avoid errors

In [18]:
ords_prods_merge=ords_prods_merge.drop(columns ='_merge')

#### Merge Customer_original & ords_prods_merge dataframes 

In [19]:
## merge dataframe by using key as "user_id"

instacart_ords_prods_customer_merge=ords_prods_merge.merge(df_customer_original, on = 'user_id', indicator=True)

In [22]:
instacart_ords_prods_customer_merge.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_from_last_order', 'product_id',
       'add_to_cart_order', 'reordered', 'Unnamed: 0', 'product_name',
       'aisle_id', 'department_id', 'prices', 'merge_indicator',
       'price_range_loc', 'busiest_day', 'busiest_days',
       'busiest_period_of_day', 'max_order', 'loyalty_flag',
       'average_spending', 'spending flag', 'frequency_of_purchase',
       'order frequency flag', 'first_name', 'last_name', 'gender', 'state',
       'age', 'date_joined', 'dependant_no', 'family_status', 'income',
       '_merge'],
      dtype='object')

## Q8: Export final data set as pkl file

In [24]:
instacart_ords_prods_customer_merge.to_pickle(r'C:\Users\inesh\Instacart Basket Analysis_01_02_2024\02 Data\Prepared Data\instacart_ords_prods_customer_merge.pkl')